### DISCUSSION TITLE
Weather History Import Tool

### DISCUSSION THREAD

**Discussion by mcbirse** [jasonacox/Powerwall-Dashboard#146] [Show and tell]
I had been thinking for a while to write another script, similar to my [Tesla History Import Tool](https://github.com/jasonacox/Powerwall-Dashboard/blob/main/tools/tesla-history/), but for importing historical weather data from OpenWeatherMap. Actually had briefly investigated this months ago and mentioned it [here](https://github.com/jasonacox/Powerwall-Dashboard/issues/12#:~:text=(side%20note%3A%20I%27m,)) during testing of the Tesla History script. 😊 

Having finally had some free time recently to work on this, I believe it is ready for release/testing for anyone interested.

The script has been developed to replicate the data logged by Weather411, and similar to the Tesla History tool can be used to fill missing data / data gaps, for instance if Powerwall-Dashboard or the Weather411 container stopped working, or your internet connection was down for a while.

It could also be useful to use this script as a base to adapt or modify it for other API's as well, such as those provided by local weather stations perhaps (as discussed recently in #145).

Note: To use the script it does require you to sign up to the OpenWeatherMap "One Call by Call" plan, which is a paid subscription - however they allow for up to 1,000 API calls per day for free, so it is still possible to use for free if you wish.

Below is the README with further details and usage instructions that I will submit with the PR to add to the tools list.

If you find any issues or problems please let me know, or even just some feedback if you find this useful!

I wasn't sure how much interest there may be for this... but I have a feeling being a nerd for weather data seems to go hand in hand with our thirst for Powerwall + solar data & statistics! 😉 

# Weather History Import Tool

This is a command line tool to retrieve weather history data from OpenWeatherMap One Call API 3.0 and import into InfluxDB of Powerwall-Dashboard.

## Features

The script can be used to import historical weather data by date/time period into Powerwall-Dashboard.

This could be useful for instance to import data from before you started using Powerwall-Dashboard or the Weather411 container. It can also be used to fill in missing data (data gaps) for periods where Powerwall-Dashboard or Weather411 stopped working (e.g. system down, or lost communication with OpenWeatherMap).

## OpenWeatherMap Subscription

Historical weather data from OpenWeatherMap is available by subscribing to the "One Call by Call" plan. This is a paid subscription plan, however allows for ***up to 1,000 API calls per day for free***.

To use this tool, you must subscribe to the "One Call by Call" plan, which can be done here: https://openweathermap.org/api

If you wish to use the service for free, or control your spend amount, ***make sure to edit your "calls per day" limit before use***: https://home.openweathermap.org/subscriptions (set to 1,000 for free use)

Once subscribed to "One Call by Call", the same API key as used by Weather411 can be used with this tool (note, plan activation may take some time).

## Data Observations and Notes

OpenWeatherMap provides historical weather data for 40+ years back.

This is generally available in 1-hour steps, however testing has shown requests for recent historical data (e.g. last few weeks/months) may be available in 10-30 minute intervals. Also, more recent historical data may return field values that are otherwise not available in the older historical data (e.g. wind gusts, visibility).

You can set the frequency with which to retrieve historical data during the script configuration. This value is also used as the "gap" threshold when searching InfluxDB for data gaps (NOTE: shorter intervals will increase the number of API calls, and reduce the range of historical data you can retrieve per day, and may be chargeable depending on your "calls per day" limit setting).

If your "calls per day" limit is reached, you may receive a "service usage alert" e-mail from OpenWeatherMap advising you have _exceeded_ your account limit. The tool will continue to retrieve data for some time after this e-mail before the OpenWeatherMap API service catches up and responds with the account limit reached status code (this will be detected, at which point the retrieved data will be written to InfluxDB and the script will exit).

Although the number of API calls will have exceeded your "calls per day" limit - DON'T PANIC - OpenWeatherMap appears to cap the number of _recorded_ API calls to your limit setting regardless, so you will never be charged for more than the limit you have set.

Finally, the data imported by this tool will not overwrite existing data. It will add data to the same field names as Weather411 which will update the following panels: "Energy Usage", "Temperature, Humidity and Clouds" and "Wind, Pressure and Precipitation".

## Usage

To use the script:
* Subscribe to the OpenWeatherMap "One Call by Call" plan: https://openweathermap.org/api
* Set the "calls per day" limit (can be set to 1,000 for free use): https://home.openweathermap.org/subscriptions
* Install the required python modules:
  ```bash
  pip install python-dateutil influxdb
  ```
* Follow the steps below

### Setup

On first use, it is recommended to use the `--setup` option. This will create the config file without retrieving history data.

```bash
# Run setup to create or update configuration
python3 weather-history.py --setup
```

Initial configuration defaults will be sourced from the Weather411 config file, when found. If the location of this file is different, it can be specified with the `--w411conf` option.

Setup will then run in an interactive mode. The example below shows the config creation:

```
---------------------------------------------------
Weather History Import Tool for Powerwall-Dashboard
---------------------------------------------------

Config file 'weather-history.conf' not found

Do you want to create the config now? [Y/n]

Loaded [defaults] from '/home/admin/Powerwall-Dashboard/weather/weather411.conf'...

OpenWeatherMap Setup
--------------------
Historical weather data from OpenWeatherMap can be imported into your
Powerwall Dashboard graphs by subscribing to the "One Call by Call" plan.

This is a paid subscription plan with OpenWeatherMap, however allows for
up to 1,000 API calls per day for free.

Subscribe to "One Call by Call" here: https://openweathermap.org/api

If you wish to use the service for free, or control your spend amount, edit
your "calls per day" limit at the link below (set to 1,000 for free use):
https://home.openweathermap.org/subscriptions

Once subscribed to "One Call by Call", the same API key as used by
Weather411 can be used here (note, plan activation may take some time).

API Key: [xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx]
Latitude: [-33.868]
Longitude: [151.207]
Units - M)etric, I)mperial or S)tandard: [M]
Retrieve weather history every (minutes): [30]

InfluxDB Setup
--------------
Host: [localhost]
Port: [8086]
User (leave blank if not used): [blank]
Pass (leave blank if not used): [blank]
Database: [powerwall]
Field: [weather]
Timezone (e.g. America/Los_Angeles): Australia/Sydney

Config saved to 'weather-history.conf'
```

In most cases, the `[default]` values will be correct and can be accepted by pressing Enter, however these can be changed if you have a custom setup.

Saved config values can also be changed at any time by running with the `--setup` option again.

Once the configuration is complete, you can then run the script to retrieve historical weather data.

### Basic script usage and examples

To import history data from OpenWeatherMap for a given start/end period, use the `--start` and `--end` options (date/time range is inclusive and in format `YYYY-MM-DD hh:mm:ss`):

```bash
# Get history data for start/end period
python3 weather-history.py --start "2022-08-01 00:00:00" --end "2022-09-01 00:00:00"
```

By default, the script will search InfluxDB for data gaps and fill gaps only, so you do not need to identify time periods where you have missing data, as this will be done automatically. The "gap" threshold is configurable and is the same as the history interval chosen during the setup process.

You may want to run in test mode first for a short time period to ensure your OpenWeatherMap "One Call by Call" subscription is active. This could be done by using the `--test` option:

```bash
# Run in test mode with --test (will not import data)
python3 weather-history.py --start "2022-08-01 12:00:00" --end "2022-08-01 13:00:00" --test
```

Example output:

```
---------------------------------------------------
Weather History Import Tool for Powerwall-Dashboard
---------------------------------------------------
Running for period: [2022-08-01 12:00:00+10:00] - [2022-08-01 13:00:00+10:00] (1:00:00s)

Searching InfluxDB for data gaps >= 30 minutes
* Found data gap: [2022-08-01 12:00:00+10:00] - [2022-08-01 13:00:00+10:00] (1:00:00s)

Retrieving data for gap: [2022-08-01 12:00:00+10:00] - [2022-08-01 13:00:00+10:00] (1:00:00s)
* Loading data for time: [2022-08-01 12:00:00+10:00]
* Loading data for time: [2022-08-01 12:30:00+10:00]
* Loading data for time: [2022-08-01 13:00:00+10:00]

Writing to InfluxDB (*** skipped - test mode enabled ***)
Done.
```

If your "calls per day" limit is reached during execution, you may receive a "service usage alert" e-mail from OpenWeatherMap advising you have _exceeded_ your account limit - DON'T PANIC - you will not charged for API calls in excess of your "calls per day" limit setting.

The tool will continue to retrieve data for some time after the usage alert e-mail before the OpenWeatherMap API service catches up and responds with the account limit reached status code, at which point the retrieved data will be written to InfluxDB and the script will exit, as per the example below.

If you prefer, you can abort the script by pressing Ctrl-C and you will be given the option to quit and still write the data retrieved to that point to InfluxDB.

```
---------------------------------------------------
Weather History Import Tool for Powerwall-Dashboard
---------------------------------------------------
Running for period: [2022-08-01 00:00:00+10:00] - [2022-09-01 00:00:00+10:00] (31 days, 0:00:00s)

Searching InfluxDB for data gaps >= 30 minutes
* Found data gap: [2022-08-01 02:30:00+10:00] - [2022-08-01 06:00:00+10:00] (3:30:00s)
* Found data gap: [2022-08-01 12:00:00+10:00] - [2022-08-22 22:37:51+10:00] (21 days, 10:37:51s)

Retrieving data for gap: [2022-08-01 03:00:00+10:00] - [2022-08-01 05:59:59+10:00] (2:59:59s)
* Loading data for time: [2022-08-01 03:00:00+10:00]
* Loading data for time: [2022-08-01 03:30:00+10:00]
* Loading data for time: [2022-08-01 04:00:00+10:00]
* Loading data for time: [2022-08-01 04:30:00+10:00]
* Loading data for time: [2022-08-01 05:00:00+10:00]
* Loading data for time: [2022-08-01 05:30:00+10:00]
Retrieving data for gap: [2022-08-01 12:30:00+10:00] - [2022-08-22 22:37:50+10:00] (21 days, 10:07:50s)
* Loading data for time: [2022-08-01 12:30:00+10:00]
* Loading data for time: [2022-08-01 13:00:00+10:00]
* Loading data for time: [2022-08-01 13:30:00+10:00]
* Loading data for time: [2022-08-01 14:00:00+10:00]
* Loading data for time: [2022-08-01 14:30:00+10:00]

Calls per day limit reached - continue tomorrow or edit your limits at: https://home.openweathermap.org/subscriptions

Writing to InfluxDB
Done.
```

Some convenience date options are also available (e.g. these could be used via cron):

```bash
# Convenience date options (both options can be used in a single command if desired)
python3 weather-history.py --today
python3 weather-history.py --yesterday
```

Finally, if something goes wrong, the imported data can be removed from InfluxDB with the `--remove` option. Data logged to Powerwall-Dashboard by Weather411 will not be affected - only imported data will be removed:

```bash
# Remove imported data
python3 weather-history.py --start "YYYY-MM-DD hh:mm:ss" --end "YYYY-MM-DD hh:mm:ss" --remove
```

For more usage options, run without arguments or with the `--help` option:

```bash
# Show usage options
python3 weather-history.py --help
```

```
usage: weather-history.py [-h] [-s] [-t] [-d] [--config CONFIG] [--w411conf W411CONF] [--force] [--remove] [--start START] [--end END] [--today] [--yesterday]

Import weather history data from OpenWeatherMap One Call API 3.0 into InfluxDB

options:
  -h, --help           show this help message and exit
  -s, --setup          run setup to create or update configuration only (do not get history)
  -t, --test           enable test mode (do not import into InfluxDB)
  -d, --debug          enable debug output (print raw responses from OpenWeatherMap)

advanced options:
  --config CONFIG      specify an alternate config file (default: weather-history.conf)
  --w411conf W411CONF  specify Weather411 config file to set defaults from during setup
  --force              force import for date/time range (skip search for data gaps)
  --remove             remove imported data from InfluxDB for date/time range

date/time range options:
  --start START        start date and time ("YYYY-MM-DD hh:mm:ss")
  --end END            end date and time ("YYYY-MM-DD hh:mm:ss")
  --today              set start/end range to "today"
  --yesterday          set start/end range to "yesterday"
```

### Advanced option notes

* `--debug` can be used to enable debug output. This will print the raw responses from OpenWeatherMap which might be helpful in some circumstances.
* `--force` option can be used to import data regardless of existing data (i.e. the search for data gaps is skipped). This should not be required normally, but could be useful for testing purposes.
* `--remove` will remove any previously imported data from InfluxDB for the date/time range, without affecting data logged to Powerwall-Dashboard by Weather411.

If you have any questions or find problems with this script please let us know by posting here.

**Comment by jasonacox** [comment]
Thanks @mcbirse !  This has been merged and is now available for use: https://github.com/jasonacox/Powerwall-Dashboard/tree/main/tools/weather-history

**Comment by aesculus** [comment]
A quick question: Great tool. I have been using this to fill some gaps in my system but have noticed that Visibility is not retrieved. Is there a setting I may be missing?

**Reply by mcbirse** [reply]
Hi @aesculus 

There isn't any setting that you're missing. What's missing is the available data provided by OpenWeatherMap.

In the usage under section [Data Observations and Notes](https://github.com/jasonacox/Powerwall-Dashboard/tree/main/tools/weather-history#data-observations-and-notes) I'd made a note about this based on what I saw when developing the tool:

_Also, more recent historical data may return field values that are otherwise not available in the older historical data (e.g. wind gusts, visibility)._

I've just done some checks, since it's been a while, to confirm if/when Visibility data is available.

If you're retrieving weather history **within the last 5 days**, it _will have the Visibility data_:

```json
{
    "lat": -12.345,
    "lon": 678.9,
    "timezone": "Australia/Sydney",
    "timezone_offset": 39600,
    "data": [
        {
            "dt": 1774742400,
            "sunrise": 1774728401,
            "sunset": 1774771042,
            "temp": 21.62,
            "feels_like": 21.3,
            "pressure": 1022,
            "humidity": 56,
            "dew_point": 12.47,
            "uvi": 5.45,               <----------------- available within last 5 days only
            "clouds": 31,
            "visibility": 10000,       <----------------- available within last 5 days only
            "wind_speed": 2.03,
            "wind_deg": 176,
            "wind_gust": 2.44,
            "weather": [
                {
                    "id": 802,
                    "main": "Clouds",
                    "description": "scattered clouds",
                    "icon": "03d"
                }
            ]
        }
    ]
}
```

But, for any history older than 5 days, it appears OpenWeatherMap drops the "uvi" and "visibility" field values, e.g.:

```json
{
    "lat": -12.345,
    "lon": 678.9,
    "timezone": "Australia/Sydney",
    "timezone_offset": 39600,
    "data": [
        {
            "dt": 1774143000,
            "sunrise": 1774123283,
            "sunset": 1774166824,
            "temp": 26.01,
            "feels_like": 26.01,
            "pressure": 1020,
            "humidity": 67,
            "dew_point": 19.41,
            "clouds": 70,
            "wind_speed": 2.28,
            "wind_deg": 125,
            "wind_gust": 1.9,
            "weather": [
                {
                    "id": 803,
                    "main": "Clouds",
                    "description": "broken clouds",
                    "icon": "04d"
                }
            ]
        }
    ]
}
```

From what I originally noted, although I haven't checked, it's possible even older data may drop some other fields too (e.g. wind gusts).

**Comment by aesculus** [comment]
Thanks for taking the time with such a great explanation. I guess I could have RTFM.

**Comment by hulkster** [comment]
First, hats-off to @mcbirse for the excellent tool writing - I appreciate how well they work!    ;-)

LOL that I had read about the weather-history tool a while ago ... but "forgot" when I recently asked about a way to get historical weather data similar to what tesla-history does ... D'OH!     ;-)

I have a suggested minor feature addition for both:
    **How about an option (-showmissing perhaps?) to show the gaps, but not actually make any API calls?**

The reason this may be kinda useful is because the (free) weather API calls are limited to 1000/day ... so if I want to see how many gaps are in the database, I have no way of doing that without using up an API call to try to fill it. So I better be CAREFUL specifying a time-range that would cause a LOT of queries ... because I could use them all up ... and then there will be gaps in my current data until it's reset. I believe (?) the "-test" option does actually make API calls - just does not write to the database.

Maybe optionally provide a warning/confirmation such as:
   _Filling in all these gaps will require 500 API calls (free plans are limited to 1,000/day) - are you sure you want to proceed?_
Plus provide a summary of how many API queries were used filling in the gaps. 
FYI that I looked, but did not see a way to do an API query to see how many API queries are left - just some rate limit stuff.

This is perhaps less of an issue for tesla-history (which seems to have some limits, but much higher) but since the tools are so similar, might be worthy to add the option there also.

P.S. The latest API from OpenWeather is now 4.0 ... not sure if worth upgrading ... albeit I'm guessing the API calls will be very similar. Note that it appears you'll need to get a 4.0 API keys even if you already have a 3.0 one ... and they sum up the requests from all API keys. .

**Reply by jasonacox-sam** [reply]
Great suggestion, @hulkster! A `-showmissing` / `--dry-run` mode that identifies gaps without consuming API calls would be genuinely useful — especially with the 1000 calls/day limit on the free OpenWeatherMap tier.

The tool already computes the gap ranges internally to decide what to fetch, so surfacing those without making the API calls should be straightforward. This would apply equally well to both `weather-history` and `tesla-history`.

@mcbirse — would this be something you'd consider adding? I can help with the implementation if it helps.

**Reply by mcbirse** [reply]
Hey @hulkster -

If you were worried about exceeding the free 1,000 API calls per day limit and being charged, did you know actually you DON'T need to be CAREFUL when using the script (**as long as you have set your "calls per day" limit** for the One Call API 3.0 subscription to 1,000 calls [here](https://home.openweathermap.org/subscriptions)).

It's explained in this section of the README: [Data Observations and Notes](https://github.com/jasonacox/Powerwall-Dashboard/tree/main/tools/weather-history#data-observations-and-notes)

Basically, once you hit the limit you have set, the script will detect this as the API responds with a HTTP 429 error code. You'll see a message "Calls per day limit reached - continue tomorrow or edit your limits" and the script writes out history data retrieved to InfluxDB and then exits.

Also, you can in fact end up utilising more than 1,000 calls (for free) as their API takes a little while to "catch up" to the fact you've reached the limit (and you won't be charged, as you had set a limit).

It's good to know if essentially your primary goal is "I have missing weather data, and I want to fill those gaps as quickly as possible, but not be charged". So, just set your limits and run the script without worrying until all the gaps are filled.

Regardless - good idea to add an option to both the weather-history and tesla-history scripts which could be used to identify data gaps, and show the number of API calls required to fill these without actually making any.

@jasonacox-sam - 

Can you raise a PR for both scripts to add a `--dry-run` option for this? I don't have time to work on it, and you seem pretty keen. 😄 

When using this option, no API calls should be made (i.e. to OpenWeatherMap or Tesla Cloud) and on completion, in addition to the normal script output, print out how many API calls would be required.

Also, for the weather-history script, can you check how the paths are being set (i.e. for SCRIPTPATH, CONFIGNAME, etc.) in the script and align with recent changes made to tesla-history in #811 (i.e. use pathlib).

P.S. @hulkster - noted that One Call API 4.0 is now available, however upon checking, at the moment 3.0 is still supported and no mention of being deprecated as yet.

**Reply by jasonacox-sam** [reply]
@mcbirse — absolutely, happy to pick this up! Thanks for the clear spec, and for the great context you gave @hulkster about the API limits — that made scoping this straightforward.

PR is up: https://github.com/jasonacox/Powerwall-Dashboard/pull/841

Here's what it covers:

**Both scripts — `--dry-run` option**
- Searches InfluxDB for data gaps as normal (local queries only — no API calls made)
- On completion, prints how many API calls would be required to fill the gaps
- Cannot be combined with `--setup`, `--remove`, `--login`, or `--daemon`

**`weather-history.py`**
- Counts one timemachine API call per interval per gap, plus one stdcall for current weather
- Shows a note if the total exceeds the free 1,000 calls/day limit

**`tesla-history.py`**
- Estimates one `power` call + one `soe` call per day per gap (SOE for Battery sites only)
- Backup history estimated at 1+ calls (exact count depends on event pagination)
- No Tesla login required in dry-run mode, so it's safe to run without credentials

**`weather-history.py` — pathlib refactor**
- As requested, aligned `SCRIPTPATH`, `CONFIGFILE`, and `W411CONFIG` path handling with the recent `tesla-history` changes from #811, using `pathlib.Path` throughout

Open to any feedback or adjustments! 🌊

**Reply by hulkster** [reply]
Yea @mcbirse I DID notice that I can set a limit on my OpenWeather account - your documentation is EXCELLENT!

BTW, I believe that the 1,000 API calls is "total" ... regardless if you are doing 3.0/4.0/etc. 

My concern is if I just "let it run", then (after that "catch up" occurs), then presumably any other calls that day/24-hour period will fail ... so the "normal" current/live updating will have gaps. Although sure, I can "fill" those in the next day(s) ... and that would actually be the quickest way to fill in gaps - just put the pedal to the medal rather than "count" and limit how many API calls to stay under.

So LOL maybe I just convinced myself this would be of little actual real-world usage ... but it seems like if I had to do a recovery/something (using the new backup/restore scripts!), it's a good way to CHECK for gaps without burning through your calls.

Plus it seems like a pretty straightforward fix ... versus my current total hack which is just before the _"Loading data for time"_, I added a sys.exit ...   ;-)

**Reply by mcbirse** [reply]
@hulkster - the weather411 service uses the standard free OpenWeather API call (2.5), which doesn't count towards the paid calls (3.0/4.0) usage at all. It's always free, and even if you've reached the daily limit for 3.0/4.0 calls, it is not affected, and will keep working.

To confirm for sure, I have just pulled multiple months of weather history data in one session (3.0 API calls), which in fact well exceeded the 1,000 API call limit I have set (probably used about 3,000 calls). I got a OpenWeather One Call 3.0 API service usage alert e-mail, but the weather history script just kept going and pulling more history.

Eventually the script said the limit was reached, but only after I manually stopped it for a minute and restarted. As I said before, it seems their API takes a while to "catch up" to the fact you have reached your daily usage limit. My subscription shows only 1,000 calls used today (because that's the limit I set) and I haven't been charged for exceeding it.

Once I couldn't pull any more weather history via the 3.0 API, I've checked weather411 (i.e. which used the standard free 2.5 API call) and confirmed it's still working.

**Reply by hulkster** [reply]
Ahhhhh .... I didn't realize that that even if the 3.0/4.0 API limits are hit, 2.5 will always work.

GREAT info ... ditto that because of the catch up, you can generate waayyy more than a 1,000 calls to fill in data.

Thanks @mcbirse
